crates/block2/src/traits.rs

```diff
@@ -135,3 +138,225 @@ impl_traits!(t0: T0, t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7: T7, t8:
+/// where `MyClass::sel` has a signature compatible with the block you want,
+/// after which the string can be slightly modified to fit an actual block
+/// instead of a method: see below.
+///
+/// A more thorough but manual approach is to only follow the rules described
+/// below.
+///
+/// [enc2s]: objc2::encode::Encoding#impl-Display-for-Encoding
+///
+/// # Encoding string generation
+///
+/// This is the result of the `@encode` Objective-C compiler directive. The
+/// [Apple documentation] and [GCC documentation] explain how each base type is
+/// encoded into a string representation. See there for a somewhat-formal
+/// specification and a few basic examples. See also [`Encoding`].
+///
+/// See also the [GCC method signatures] section. It is mostly valid for blocks
+/// as well, since they are basically functions with captured environment --
+/// i.e. closures, except that no selector is implicitely sent, only the block
+/// object is. In short, the "signature" is a null-terminated string, composed
+/// of the following, in order:
+///
+///  * The return type, including type qualifiers. For example, a block
+///    returning `int` ([`i32`]) would have `i` here.
+///  * The total size (in bytes) required to pass all the parameters: the call
+///    frame size. This includes the hidden block object parameter that is
+///    passed as a pointer, so at least 4 bytes when under a 32-bit system or
+///    most probably 8 bytes when under a 64-bit one.
+///  * Each argument, with the type encoding, followed by the offset (in bytes)
+///    of the argument in the list of parameters. The first is the always the
+///    hidden block object pointer and is therefore `@?0`.
+///
+/// Examples:
+///
+/// | Objective-C signature    | Runtime encoding                           |
+/// | ------------------------ | ------------------------------------------ |
+/// | `void (^)(void)`         | `v8@?0`                                    |
+/// | `int (^)(void)`          | `i8@?0`                                    |
+/// | `int (^)(float)`         | `i12@?0f8`                                 |
+/// | `int (^)(float, _Bool)`  | `i16@?0f8B12`                              |
+/// | `void (^)(int*)`         | `v16@?0^i8`                                |
+/// | `void (^)(NSError*)`     | `v16@?0@8` or `v16@?0@"NSError"8`          |
+/// | `NSError* (^)(NSError*)` | `@16@?0@8` or `@"NSError"16@?0@"NSError"8` |
+///
+/// [Apple documentation]: https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ObjCRuntimeGuide/Articles/ocrtTypeEncodings.html
+/// [`Encoding`]: objc2::encode::Encoding
+/// [GCC documentation]: https://gcc.gnu.org/onlinedocs/gcc/Type-encoding.html
+/// [GCC method signatures]: https://gcc.gnu.org/onlinedocs/gcc/Method-signatures.html
+pub unsafe trait ManualBlockEncoding {
+    /// The function's input argument types.
+    type Arguments: EncodeArguments;
+    /// The function's return type.
+    type Return: EncodeReturn;
```

Comment:
Did you consider not having these two parameters here? I can see how they're somewhat beneficial for ensuring that you don't accidentally pass one block encoding helper to another place where it wasn't intended, but it also makes things more cumbersome for the user than strictly necessary (and we already have the debug assertion, so it shouldn't be that big of an issue).

Reply:
Yes, I did, but I simply followed [your comment just above](https://github.com/madsmtm/objc2/pull/636#issuecomment-2227388348) as I thought it meant you wanted it this way specifically in order to force users to explicitly think about what they are typing.

Reply:
Heh, I just sorta sketched it out back then. I think now that I see it that I have a slight preference towards not having these, but no strong preference, I'll leave the decision to you.

Reply:
I see what you mean. However, having these dummy types did help me even when simply testing: copy-pasting without appropriately changing the types made the compilation failed and therefore made me realize "oh yes, I forgot to update this". Also, when refactoring, it is easy to change the input or output type of a closure without it being written explicitly, so having a compile-time check helps in this regard as well.

All in all, I think it wasn't such a bad idea from the start if unsafety can truly arise from specifying a wrong signature string. Testing myself with the `NetworkExtension`'s only part I found requiring a signed block, but specifying a completely wrong string made no difference whatsoever if I remember correctly, so it might not be that important in the end. If safety looks to be more of a guess than a true requirement, then I'll remove the types; otherwise, I feel like keeping them is appropriate, but I might have to take a look at some examples again to see the difference.

Reply:
> but specifying a completely wrong string made no difference whatsoever if I remember correctly, so it might not be that important in the end

See https://github.com/madsmtm/objc2/issues/442#issuecomment-2284932726 - I can imagine that specifying a random string in certain cases doesn't make a difference if you have a block with only one argument (since it's most likely going to place things correctly then), but for multi-argument blocks you're likely to run into UB. In any case, since it's invoked via. `NSInvocation`, it's definitely required for soundness.

(I posted it on the issue because I might want to reference the information later on)

Reply:
Ah, thanks for all the details! It might even inspire me to add more to the added safety comments in order to justify the unsafety and relieve users of potential doubts similar to what I had.

Regarding the question of whether to keep the dummy generic types associated to the smuggling trait, I've thought about it more since then and found that it could serve as self-documentation and help for readability in addition to what I had already mentioned: reading the signature string is pretty much impossible when not initiated; having the types explicitly written should help make things clearer and therefore avoid oversights and general fat-finger-ness; also, since this API is already surprising to use due to the smuggling trait, I think adding this much on top should be OK.

If it's alright with you, I think I'll prefer keeping it like this. If I haven't already, I think I'll add a mention of this in the documentation somewhere so users may understand better why something that looks useless at first glance is required.